vcls/magento2/default.vcl

```diff
@@ -0,0 +1,229 @@
+    }
+
+    if (req.method != "GET" &&
+        req.method != "HEAD" &&
+        req.method != "PUT" &&
+        req.method != "POST" &&
+        req.method != "TRACE" &&
+        req.method != "OPTIONS" &&
+        req.method != "DELETE") {
+          /* Non-RFC2616 or CONNECT which is weird. */
+          return (pipe);
+    }
+
+    # We only deal with GET and HEAD by default
+    if (req.method != "GET" && req.method != "HEAD") {
+        return (pass);
+    }
+
+    # Bypass customer, shopping cart, checkout
+    if (req.url ~ "/customer" || req.url ~ "/checkout") {
+        return (pass);
+    }
+
+    # Bypass health check requests
+    if (req.url ~ "^/(pub/)?(health_check.php)$") {
+        return (pass);
+    }
+
+    # if the backend is healthy, limit the grace
+    if (std.healthy(req.backend_hint)) {
+        set req.grace = /* {{ grace_period }} */s;
+        set req.http.grace = "normal (healthy server)";
+    } else {
+        set req.http.grace = "unlimited (unhealthy server)";
+    }
+
+    set req.http.grace = "none";
+
+    # normalize url in case of leading HTTP scheme and domain
+    set req.url = regsub(req.url, "^http[s]?://", "");
+
+    # collect all cookies
+    std.collect(req.http.Cookie);
+
+    # Remove all marketing get parameters to minimize the cache objects
+    if (req.url ~ "(\?|&)(gclid|cx|ie|cof|siteurl|zanpid|origin|fbclid|mc_[a-z]+|utm_[a-z]+|_bta_[a-z]+)=") {
+        set req.url = regsuball(req.url, "(gclid|cx|ie|cof|siteurl|zanpid|origin|fbclid|mc_[a-z]+|utm_[a-z]+|_bta_[a-z]+)=[-_A-z0-9+()%.]+&?", "");
+        set req.url = regsub(req.url, "[?|&]+$", "");
+    }
+
+    # Static files caching
+    if (req.url ~ "^/(pub/)?(media|static)/") {
+        # Static files should not be cached by default
+        return (pass);
+
+        # But if you use a few locales and don't use CDN you can enable caching static files by commenting previous line (#return (pass);) and uncommenting next 3 lines
+        #unset req.http.Https;
+        #unset req.http./* {{ ssl_offloaded_header }} */;
+        #unset req.http.Cookie;
+    }
+
+    # Authenticated GraphQL requests should not be cached by default
+    if (req.url ~ "/graphql" && req.http.Authorization ~ "^Bearer") {
+        return (pass);
+    }
+
+    return (hash);
+}
+
+sub vcl_hash {
+    if (req.http.cookie ~ "X-Magento-Vary=") {
+        hash_data(regsub(req.http.cookie, "^.*?X-Magento-Vary=([^;]+);*.*$", "\1"));
+    } else {
+    	hash_data("");
+    }
+
+    hash_data(req.http./* {{ ssl_offloaded_header }} */);
+    /* {{ design_exceptions_code }} */
+
+    if (req.url ~ "/graphql") {
+        hash_data(req.http.Store);
+        hash_data(req.http.Content-Currency);
+    }
+}
+
+sub vcl_backend_response {
+    # tag the object to purge it later
+    ykey.add_header(beresp.http.X-Magento-Tags);
+    ykey.namespace("pool");
+    ykey.add_header(beresp.http.X-Pool);
+    ykey.namespace_reset();
+
+    set beresp.grace = 3d;
```

Comment:
Move this to a 3D stale-on-error. Use the grace value in vcl_recv here.

Reply:
PR welcome